File: src/main/java/org/hoohoot/homelab/manager/notifications/parser/ParseSeries.java
package org.hoohoot.homelab.manager.notifications.parser;

import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.Optional;

@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
public class ParseSeries extends ArrParser {
    private final Optional<JsonObject> series;
    private final Optional<JsonArray> episodes;
    private final Optional<JsonArray> episodeFiles;
    private final Optional<JsonObject> release;
    private final Optional<String> downloadClient;

    private static final String DEFAULT_VALUE = "unknown";

    private ParseSeries(JsonObject payload) {
        super(Optional.ofNullable(payload.getJsonObject("series")).map(series -> series.getJsonArray("tags")));
        this.series = Optional.ofNullable(payload.getJsonObject("series"));
        this.episodes = Optional.ofNullable(payload.getJsonArray("episodes"));
        this.episodeFiles = Optional.ofNullable(payload.getJsonArray("episodeFiles"));
        this.release = Optional.ofNullable(payload.getJsonObject("release"));
        this.downloadClient = Optional.ofNullable(payload.getString("downloadClient"));
    }

    public static ParseSeries from(JsonObject payload) {
        return new ParseSeries(payload);
    }

    public String quality() {
        return this.episodeFiles.map(jsonArray -> jsonArray.getJsonObject(0))
                .stream().findFirst()
                .map(e -> e.getString("quality"))
                .orElse(DEFAULT_VALUE);
    }

    public String seriesName() {
        return this.series.map(json -> json.getString("title"))
                .orElse(DEFAULT_VALUE);
    }

    public String seasonAndEpisodeNumber() {
        return this.episodes.map(jsonArray -> jsonArray.getJsonObject(0))
                .map(e -> {
                    var episodeNumber = e.getInteger("episodeNumber");
                    var seasonNumber = e.getInteger("seasonNumber");

                    return "S%02dE%02d".formatted(seasonNumber, episodeNumber);
                }).orElse(DEFAULT_VALUE);
    }

    public String episodeName() {
        return this.episodes.map(jsonArray -> jsonArray.getJsonObject(0))
                .map(episode -> episode.getString("title"))
                .orElse(DEFAULT_VALUE);
    }

    public String downloadClient() {
        return this.downloadClient.orElse(DEFAULT_VALUE);
    }

    public String indexer() {
        return release
                .map(r -> r.getString("indexer").replace(" (Prowlarr)", ""))
                .orElse(DEFAULT_VALUE);
    }
}

File: src/main/kotlin/org/hoohoot/homelab/manager/notifications/parser/ParseSeries.kt
package org.hoohoot.homelab.manager.notifications.parser

import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject

private const val DEFAULT_VALUE = "unknown"

class ParseSeries private constructor(payload: JsonObject) {
    private val series: JsonObject? = payload.getJsonObject("series")
    private val episodes: JsonArray? = payload.getJsonArray("episodes")
    private val episodeFiles: JsonArray? = payload.getJsonArray("episodeFiles")
    private val release: JsonObject? = payload.getJsonObject("release")
    private val downloadClient: String? = payload.getString("downloadClient")

    companion object {
        @JvmStatic
        fun from(payload: JsonObject) = ParseSeries(payload)
    }

    fun quality(): String = episodeFiles
        ?.getJsonObject(0)
        ?.getString("quality")
        ?: DEFAULT_VALUE

    fun seriesName(): String = series
        ?.getString("title")
        ?: DEFAULT_VALUE

    fun seasonAndEpisodeNumber(): String = episodes
        ?.getJsonObject(0)
        ?.let {
            val episodeNumber = it.getInteger("episodeNumber")
            val seasonNumber = it.getInteger("seasonNumber")
            "S%02dE%02d".format(seasonNumber, episodeNumber)
        }
        ?: DEFAULT_VALUE

    fun episodeName(): String = episodes
        ?.getJsonObject(0)
        ?.getString("title")
        ?: DEFAULT_VALUE

    fun downloadClient(): String = downloadClient ?: DEFAULT_VALUE

    fun indexer(): String = release
        ?.getString("indexer")
        ?.replace(" (Prowlarr)", "")
        ?: DEFAULT_VALUE

    fun imdbLink(): String = series
        ?.getString("imdbId")
        ?.let { "https://www.imdb.com/title/$it/" }
        ?: DEFAULT_VALUE

    fun requester(): String = series?.getJsonArray("tags")
        ?.requester()
        ?: DEFAULT_VALUE
}